[V1][Metrics] Allow V1 AsyncLLM to use custom logger

liuzijing2014 opened the pull request:
### Content

Update V1 AsynLLM to take in custom logger that extends `StatLoggerBase`. This would help to achieve function on parity as V0 regarding customization support for logging.

Follow V0 style on customizing loggers:
- Take in optional `stat_loggers` from `init`.
- Add `add_logger` and `remove_logger` functions.

### Test Plan

```
pytest tests/entrypoints/openai/test_metrics.py 
...
======================== 13 passed, 1 skipped in 207.14s (0:03:27) ========================


pytest tests/v1/engine/test_async_llm.py
...
======================= 10 passed, 10 warnings in 160.87s (0:02:40) =======================
```

liuzijing2014 committed:
[V1][Metrics] Allow AsyncLLM engine to use custom stat logger

Signed-off-by: Zijing Liu <liuzijing2014@gmail.com>
Signed-off-by: Mark McLoughlin <markmc@redhat.com>
Signed-off-by: Zijing Liu <liuzijing2014@gmail.com>

markmc commented on vllm/third_party/pynvml.py:
```diff
@@ -30,16 +30,16 @@
+import os
+import string
+import sys
+import threading
```
Remove this change from the PR

liuzijing2014 replied:
This is the auto format for imports ordering from `isort`. Is this formatting not desired?

njhill replied:
@liuzijing2014 the formatting change is to an unrelated file so shouldn't be included in the same PR.

markmc replied:
No, it is not desired

This code is copied from [nvidia-ml-py](https://pypi.org/project/nvidia-ml-py/) and as per #12977:

> If there are bugfixes in nvidia-ml-py in the future, we can periodically sync the code.

So, we would prefer not to modify this code from the original

We do use `isort`, but only using `pre-commit` and in `.pre-commit-config.yaml` we exclude this directory:

```
exclude: 'vllm/third_party/.*'
```

So, for example, if I checkout your branch and do:

```
$ pre-commit run --all-files isort
```

this file is not modified because of the `exclude`

Hope that helps!

markmc commented on vllm/v1/engine/async_llm.py:
```diff
@@ -56,11 +59,16 @@ def __init__(
+                    self.stat_loggers[STANDARD_LOGGING_LOGGER_NAME] = (
+                        LoggingStatLogger())
+                self.stat_loggers[PROMETHEUS_LOGGING_LOGGER_NAME] = (
+                    PrometheusStatLogger(vllm_config))
```
How about we move all of this into the loggers module so that we just do e.g.

```
    self.stat_loggers = loggers.setup(vllm_config, stat_loggers)
```

markmc reviewed (changes requested):
Ok, this quite closely matches the V0 interface, but not quite - e.g. the `record()` method is new, and we're using new classes like `IterationStats`, and `SchedulerStats`

Personally, I'm not a fan of having all of this be part of the public API in such an ad-hoc way ... the logger behaviors and the stats classes etc. are all liable to change in future IMO, breaking users of the API

However, if this is something we want to retain from V0, then this PR looks reasonable

njhill commented:
Agree with @markmc these interfaces / data structures may still be in flux so we may want to be careful exposing as part of an external API at this stage. If we do it should be clearly documented as unstable.

liuzijing2014 commented:
> Ok, this quite closely matches the V0 interface, but not quite - e.g. the `record()` method is new, and we're using new classes like `IterationStats`, and `SchedulerStats`
> 
> Personally, I'm not a fan of having all of this be part of the public API in such an ad-hoc way ... the logger behaviors and the stats classes etc. are all liable to change in future IMO, breaking users of the API
> 
> However, if this is something we want to retain from V0, then this PR looks reasonable

The main motivation is to allow users to basic logging parity between V0 and V1 ([v1 feedback thread](https://github.com/vllm-project/vllm/issues/12568#issuecomment-2702174813) ). We have some customized logging backend that we want to log stats to, and switching to V1 breaks the stats support for us. 

@njhill Proper documentation sounds reasonable. Besides inline code comments, is there any other place that you think we need to call this part out?

njhill commented on vllm/v1/engine/async_llm.py:
```diff
@@ -56,11 +56,14 @@ def __init__(
+        self.stat_loggers: dict[str, StatLoggerBase] = dict()
         if self.log_stats:
-            if logger.isEnabledFor(logging.INFO):
-                self.stat_loggers.append(LoggingStatLogger())
-            self.stat_loggers.append(PrometheusStatLogger(vllm_config))
+            if stat_loggers is not None:
+                self.stat_loggers = stat_loggers
+            else:
+                setup_default_loggers(vllm_config,
+                                      logger.isEnabledFor(logging.INFO),
+                                      self.stat_loggers)
```
Suggest moving all of this logic into the setup function:
```python
self.stat_loggers = setup_default_loggers(vllm_config, log_stats, stat_loggers)
```

Also:
- No need to pass in `logger.isEnabledFor(logging.INFO)`, that can be done inside the function
- It would probably be cleaner to eliminate `self.log_stats`. Empty `self.stat_loggers` can serve that purpose (or possibly `None`)

njhill commented on vllm/v1/metrics/loggers.py:
```diff
@@ -428,3 +431,13 @@ def build_cudagraph_buckets(vllm_config: VllmConfig) -> list[int]:
+        PrometheusStatLogger(vllm_config))
+
+    if enable_stats_log:
+        stat_loggers[STANDARD_LOGGING_LOGGER_NAME] = (LoggingStatLogger())
```
```suggestion
        stat_loggers[STANDARD_LOGGING_LOGGER_NAME] = LoggingStatLogger()
```

njhill commented on vllm/v1/metrics/loggers.py:
```diff
@@ -17,6 +17,9 @@
 
 _LOCAL_LOGGING_INTERVAL_SEC = 5.0
 
+STANDARD_LOGGING_LOGGER_NAME = "logging"
+PROMETHEUS_LOGGING_LOGGER_NAME = "prometheus"
```
nit: unnecessarily long names
```suggestion
STANDARD_LOGGER_NAME = "logging"
PROMETHEUS_LOGGER_NAME = "prometheus"
```

njhill commented on vllm/v1/engine/async_llm.py:
```diff
@@ -351,9 +356,27 @@ async def do_log_stats(
+        for stat_logger in self.stat_loggers.values():
             stat_logger.log()
 
+    def add_logger(self, logger_name: str, logger: StatLoggerBase) -> None:
+        if not self.log_stats:
+            raise RuntimeError(
+                "Stat logging is disabled. Set `disable_log_stats=False` "
+                "argument to enable.")
+        if logger_name in self.stat_loggers:
+            raise KeyError(f"Logger with name {logger_name} already exists.")
+        self.stat_loggers[logger_name] = logger
+
+    def remove_logger(self, logger_name: str) -> None:
+        if not self.log_stats:
+            raise RuntimeError(
+                "Stat logging is disabled. Set `disable_log_stats=False` "
+                "argument to enable.")
+        if logger_name not in self.stat_loggers:
+            raise KeyError(f"Logger with name {logger_name} does not exist.")
+        del self.stat_loggers[logger_name]
+
```
Are these methods really necessary if you can pass in the dict?

liuzijing2014 replied:
I add them here to keep the interface on parity as V0. Either way could work.

njhill replied:
My preference would be to exclude them for now, since the API is different to V0 anyhow. We're trying to avoid as much unnecessary code in V1 as possible.

markmc commented:
See liuzijing2014/vllm#1 for some changes to finish this PR up. Hope that helps!

Commits at https://github.com/markmc/vllm/commits/5caa8fbfb00e3a23f4d461fac2f26e059c5efd7e

markmc commented:
> Agree with @markmc these interfaces / data structures may still be in flux so we may want to be careful exposing as part of an external API at this stage. If we do it should be clearly documented as unstable.

In https://github.com/liuzijing2014/vllm/pull/1 I added:

```
class BuiltinLoggerName(str, Enum):
    """Internal names for built-in metric loggers                                                                                                                                                           
                                                                                                                                                                                                            
    These values are intended for internal use only. Accessing to                                                                                                                                           
    these built-in loggers is not considered to be part of the                                                                                                                                              
    public API.                                                                                                                                                                                             
    """
    LOGGING = "logging"
    PROMETHEUS = "prometheus"


class StatLoggerBase(ABC):
    """Interface for logging metrics.                                                                                                                                                                       
                                                                                                                                                                                                            
    API users may define custom loggers that implement this interface.                                                                                                                                      
    However, note that the `SchedulerStats` and `IterationStats` classes                                                                                                                                    
    are not considered stable interfaces and may change in future versions.                                                                                                                                 
    """
```

njhill reviewed (commented):
> @njhill Proper documentation sounds reasonable. Besides inline code comments, is there any other place that you think we need to call this part out?

I don't think so. The main place is in the docstring for the new `stat_loggers` arg to `AsyncLLM` (and also the add/remove methods but I would prefer to just omit those for now).

njhill commented on vllm/v1/engine/async_llm.py:
```diff
@@ -138,6 +136,7 @@ def from_engine_args(
         engine_args: AsyncEngineArgs,
         start_engine_loop: bool = True,
         usage_context: UsageContext = UsageContext.ENGINE_CONTEXT,
+        stat_loggers: Optional[dict[str, StatLoggerBase]] = None,
```
@liuzijing2014 could you also add something in the docstring here mentioning that the `stat_loggers` arg is unstable. I know it's documented on those data classes now but wouldn't harm to make it even clearer :)

njhill commented on vllm/v1/engine/async_llm.py:
```diff
@@ -63,11 +63,9 @@ def __init__(
-                self.stat_loggers.append(LoggingStatLogger())
-            self.stat_loggers.append(PrometheusStatLogger(vllm_config))
+        self.stat_loggers = setup_default_loggers(
+            vllm_config, self.log_stats, logger.isEnabledFor(logging.INFO),
```
Move the `logger.isEnabledFor(logging.INFO)` check inside the method?

njhill reviewed (commented):
Thanks @liuzijing2014 it looks good now, just have a couple more minor comments.

njhill reviewed (approved):
Thanks @liuzijing2014

njhill commented:
@liuzijing2014 I had one remaining comment https://github.com/vllm-project/vllm/pull/14661#discussion_r2003818156.

This also now needs rebasing - we now may create multiple sets of stats logger instances, one per engine.

liuzijing2014 commented:
@njhill Rebased and addressed the last comment. Sry for the delay.

markmc commented on tests/v1/engine/test_async_llm.py:
```diff
@@ -188,24 +203,31 @@ async def test_abort(monkeypatch: pytest.MonkeyPatch,
+            temperature=1.0,
+            seed=33,
+            n=n,
+        )
```
Everything above here in this file seems to be reformatting existing code for no reason - please revert all of these changes, thanks

markmc commented on tests/v1/engine/test_async_llm.py:
```diff
@@ -216,3 +238,38 @@ async def test_finished_flag(monkeypatch: pytest.MonkeyPatch, n: int,
+
+@pytest.mark.parametrize(
+    "loggers",
+    [[get_customized_logger_mock()]],
```
Were you intending to make this:

```
 [None, [get_customized_logger_mock()]]
```

?

liuzijing2014 commented on tests/v1/engine/test_async_llm.py:
```diff
@@ -216,3 +238,38 @@ async def test_finished_flag(monkeypatch: pytest.MonkeyPatch, n: int,
+        after.callback(engine.shutdown)
+
+        await engine.do_log_stats()
+        for loggers in engine.stat_loggers:
```
Yeah, the first loop goes through all the lists for each engine, and then the second loop goes through all loggers within each list.

markmc replied:
Sorry, I missed that!

markmc commented on vllm/v1/metrics/loggers.py:
```diff
@@ -467,3 +474,28 @@ def build_cudagraph_buckets(vllm_config: VllmConfig) -> list[int]:
+    stat_loggers: list[list[StatLoggerBase]] = []
+    for i in range(engine_num):
+        if custom_stat_loggers is not None:
+            stat_loggers.append(custom_stat_loggers)
```
Hmm, I've just realized that this custom logging API isn't really going to be compatible with the multi-engine stuff

With the standard loggers, we have an instance of the logger per engine - see below

In the case of these custom loggers, there is only one instance and the same instance will be called for all engines, and then engine ID isn't supplied to the `record()` method

To follow the internal API, a custom logger should be supplied as a class (`StatLoggerBase` subclass) which can take an engine index as a constructor argument

Would you agree @njhill ?

njhill replied:
@markmc apologies I missed this somehow. Yes you're right, it needs to be a list of StatLogger factories rather than the loggers themselves.

alec-flowers replied:
So custom_stat_loggers = Optional[list[StatLoggerFactory]]

and then each StatLoggerFactory produces unique StatLoggerBase subclass objects. 

and we would do
StatLoggerFactory.get_stat_logger(engine_index=i) -> StatLoggerBase or something of that sort?

liuzijing2014 replied:
Or could we make a deep copy of each logger object directly for each engine idx?

alec-flowers commented on vllm/v1/engine/async_llm.py:
```diff
@@ -53,7 +51,28 @@ def __init__(
+            use_cached_outputs (bool): Whether to use cached outputs.
+            log_requests (bool): Whether to log requests.
+            start_engine_loop (bool): Whether to start the engine loop.
+            stat_loggers (dict[str, StatLoggerBase]): customized stat loggers
```
This docstring here is incorrect - (dict[str, StatLoggerBase]) -> Optional[list[StatLoggerBase]]

alec-flowers replied:
Maybe adding types here is overkill? Its already clearly listed above

alec-flowers replied:
Also stat_loggers mirrors internal name self.stat_loggers, what about external_stat_loggers?

alec-flowers commented:
Thanks for this work @liuzijing2014! I am relying on this PR getting merged for some work that I am doing. Let me know if there is any way I can help you out here.

markmc commented:
I have pushed a rebase and the changes discussed above to https://github.com/markmc/vllm/tree/liuzijing2014-v1-custom-logger

markmc commented:
> I have pushed a rebase and the changes discussed above to https://github.com/markmc/vllm/tree/liuzijing2014-v1-custom-logger

To push, you can do:

```
$ git remote add -f markmc https://github.com/markmc/vllm.git
$ git push origin +markmc/liuzijing2014-v1-custom-logger:v1-custom-logger
```

njhill commented:
@liuzijing2014 are you ok if we push some last changes to get this finished?

liuzijing2014 commented:
Hi @njhill @markmc totally, sorry for acting slow on this PR. Got occupied with a few Llama4 work in the past few weeks. If you already have the changes, please go ahead and push. If not, I can grab some time this week and try to close this by Wednesday. Let me know.

njhill commented on vllm/v1/engine/async_llm.py:
```diff
@@ -52,7 +50,28 @@ def __init__(
         use_cached_outputs: bool = False,
         log_requests: bool = True,
         start_engine_loop: bool = True,
+        stat_loggers: Optional[list[type[StatLoggerBase]]] = None,
```
@markmc I wonder whether we should change `type[StatLoggerBase]` to `Callable[[VllmConfig, int], StatLoggerBase]`, and could define a shorter type name for that like
```python
StatLoggerFactory = Callable[[VllmConfig, int], StatLoggerBase]
```

which is more general but also captures the args - you could still then pass a subclass of `StatLoggerBase` which has the appropriate constructor?

Of course still making clear that this isn't yet a stable interface.

markmc replied:
Nice idea; you can pull another commit from the same branch with that change :+1:

markmc committed:
[V1][Metrics] Define StatLoggerFactory type

More general than passing sub-class types, requiring sub-class
constructors to conform to a protocol.

Signed-off-by: Mark McLoughlin <markmc@redhat.com>
Signed-off-by: Zijing Liu <liuzijing2014@gmail.com>

markmc commented on vllm/v1/metrics/loggers.py:
```diff
@@ -479,3 +492,33 @@ def build_cudagraph_buckets(vllm_config: VllmConfig) -> list[int]:
+    if not log_stats:
+        return []
+
+    def logging_factory(vllm_config: VllmConfig, engine_num: int):
+        return (LoggingStatLogger(vllm_config, engine_num)
+                if logger.isEnabledFor(logging.INFO) else None)
+
+    factories: list[StatLoggerFactory] = [
+        PrometheusStatLogger, logging_factory
+    ]
+    if custom_stat_loggers is not None:
+        factories = custom_stat_loggers
```
Yes! I actually did the `if logger.isEnabledFor()` thing via a factory, because I realized that supporting a None return from factories might be useful ... and then I forgot to add the `None` handling! Doh!

Well spotted, your version is fine by me, thanks :+1:

njhill committed:
Update vllm/v1/metrics/loggers.py

Signed-off-by: Zijing Liu <liuzijing2014@gmail.com>

markmc commented:
@njhill FYI

```
Error: vllm/v1/metrics/loggers.py:510: error: Name "factories" already defined on line 508  [no-redef]
```

njhill committed:
Fix typing

Signed-off-by: Nick Hill <nhill@redhat.com>
Signed-off-by: Zijing Liu <liuzijing2014@gmail.com>

njhill reviewed (approved):
@markmc @liuzijing2014 I think this is ready now?

liuzijing2014 commented:
LGTM, thanks for all the help! @njhill @markmc
